src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsLanguageDebugInfo.cs

```diff
@@ -194,42 +203,49 @@ public int ResolveName(string pszName, uint dwFlags, out IVsEnumDebugName? ppNam
                         {
-                            var solution = _languageService.Workspace.CurrentSolution;
-
-                            // NOTE(cyrusn): We have to wait here because the debuggers' ResolveName
-                            // call is synchronous.  In the future it would be nice to make it async.
-                            if (_breakpointService != null)
+                            var cancellationToken = waitContext.UserCancellationToken;
+                            if (dwFlags == (uint)RESOLVENAMEFLAGS.RNF_BREAKPOINT)
                             {
-                                var breakpoints = _breakpointService.ResolveBreakpointsAsync(solution, pszName, cancellationToken).WaitAndGetResult(cancellationToken);
-                                var debugNames = breakpoints.Select(bp => CreateDebugName(bp, solution, cancellationToken)).WhereNotNull().ToList();
+                                var solution = _languageService.Workspace.CurrentSolution;
 
-                                enumName = new VsEnumDebugName(debugNames);
+                                // NOTE(cyrusn): We have to wait here because the debuggers' ResolveName
+                                // call is synchronous.  In the future it would be nice to make it async.
+                                if (_breakpointService != null)
+                                {
+                                    var breakpoints = await _breakpointService.ResolveBreakpointsAsync(
+                                        solution, pszName, cancellationToken).ConfigureAwait(false);
+                                    var debugNames = await breakpoints.SelectAsArrayAsync(
+                                        async bp => await CreateDebugNameAsync(
+                                            bp, solution, cancellationToken).ConfigureAwait(false)).ConfigureAwait(false);
+
+                                    enumName = new VsEnumDebugName(debugNames);
+                                }
                             }
-                        }
+                        });
                     });
 
                     ppNames = enumName;
                     return ppNames != null ? VSConstants.S_OK : VSConstants.E_NOTIMPL;
                 }
             }
 
-            private IVsDebugName CreateDebugName(BreakpointResolutionResult breakpoint, Solution solution, CancellationToken cancellationToken)
+            private async Task<IVsDebugName> CreateDebugNameAsync(
+                BreakpointResolutionResult breakpoint, Solution solution, CancellationToken cancellationToken)
             {
                 var document = breakpoint.Document;
                 var filePath = _languageService.Workspace.GetFilePath(document.Id);
-                var text = document.GetTextSynchronously(cancellationToken);
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Is the assumption here this almost always completes synchronously? Because otherwise the ConfigureAwait(false) worries me a bit because if it doesn't complete, then we're going to be waiting on the thread pool while blocking the UI when we could have just used the UI thread originally.

Reply:
yes... that's a good point.  bleag... this is a rough area.